src/cpp/session/modules/SessionRmdNotebook.R

```diff
@@ -205,165 +185,122 @@ assign(".rs.notebookVersion", envir = .rs.toolsEnv(), "1.0")
+                  
+                  bodyEl <- .rs.extractHTMLBodyElement(val)
+                  
+                  rendered <- htmltools::htmlPreserve(bodyEl)
+                  annotated <- .rs.rnb.htmlAnnotatedOutput(rendered, "htmlwidget", meta = jsonContents)
+                  output <- knitr::asis_output(annotated, meta = jsonContents)
+                  return(output)
+               }
+               
+               # TODO: shouldn't get here?
+               NULL
+            })
+            
+            list(
+               structure(list(src = NULL), class = "source"),
+               htmlOutput
+            )
          }
-      })
       
-      # insert into document
-      injection <- paste(unlist(htmlList), collapse = "\n")
+      )
       
-      html[[i]] <- paste(injection, sep = "\n", collapse = "\n")
-      chunkIdx <- chunkIdx + 1
    }
    
-   # Inject JSON dependency information into document
-   # TODO: Resolve duplicates
-   htmlInjection <- unlist(lapply(htmlDependencies, function(dep) {
-      
-      injection <- c()
-      
-      # Inject a CSS resource
-      if (!is.null(dep$stylesheet)) {
-         cssPath <- file.path(dep$src$file, dep$stylesheet)
-         contents <- .rs.readFile(cssPath, binary = TRUE)
-         encoded <- .rs.base64encode(contents)
-         htmlAttributes <- list(
-            "data-rnb-id" = sprintf("lib/%s-%s/%s", dep$name, dep$version, dep$stylesheet),
-            "href" = sprintf("data:text/css;charset=utf8;base64,%s", encoded),
-            "rel" = "stylesheet",
-            "type" = "text/css"
-         )
-         cssHtml <- sprintf("<link %s />", .rs.listToHtmlAttributes(htmlAttributes))
-         injection <- c(injection, cssHtml)
-      }
-      
-      # Inject a JS resource
-      if (!is.null(dep$script))
-      {
-         jsPath <- file.path(dep$src$file, dep$script)
-         contents <- .rs.readFile(jsPath, binary = TRUE)
-         encoded <- .rs.base64encode(contents)
-         htmlAttributes <- list(
-            "data-rnb-id" = sprintf("lib/%s-%s/%s", dep$name, dep$version, dep$script),
-            "src" = sprintf("data:application/x-javascript;base64,%s", encoded)
-         )
-         scriptHtml <- sprintf("<script %s></script>", .rs.listToHtmlAttributes(htmlAttributes))
-         injection <- c(injection, scriptHtml)
-      }
-      
-      # Inject other header contents
-      rendered <- htmltools:::renderTags(dep)
-      headContent <- as.character(rendered$head)
-      if (length(headContent) && any(nzchar(headContent)))
-         injection <- c(injection, headContent)
-      
-      paste(injection, collapse = "\n")
-   }))
+   exit <- function(...) {
+      knitr::knit_hooks$set(knitHooks)
+   }
    
-   bodyIdx <- tail(grep("^\\s*</head>\\s*$", html, perl = TRUE), n = 1)
-   html <- c(
-      html[1:(bodyIdx - 1)],
-      htmlInjection,
-      html[bodyIdx:length(html)]
-   )
+   suppressMessages(trace(
```

Comment:
That's right; I want to add some behavior in addition to the hooks that are set by default with `knitr::render_markdown()` -- since custom hooks added to the render function override any behavior from the `knitr::render_markdown()` hooks, I think I need to copy the functionality from hooks generated by `render_markdown()` into the hooks I want, or find some other mechanism for 'layering' hooks.